protocol/x/clob/memclob/memclob.go

```diff
@@ -2003,8 +2003,14 @@ func (m *MemClobPriceTimePriority) RemoveOrderIfFilled(
 	// Get current fill amount for this order.
 	exists, orderStateFillAmount, _ := m.clobKeeper.GetOrderFillAmount(ctx, orderId)
 
-	// If there is no fill amount for this order, return early.
+	// If there is no fill amount for this order, error log and return early. Note this is an
+	// error since this function should only be called for orders that were partially or fully filled.
 	if !exists {
+		m.clobKeeper.Logger(ctx).Error(
```

Comment:
These orders were filled in the last block so they should always have a non-zero fill amount in state

Reply:
Do all the callsites for this function only contain orders filled in last block?

Reply:
nvm, mixed up fn names. I see that this has only one callsite populating the `filledOrderIds` from PPME.

Reply:
In the future do we want to panic in this situation instead of early returning? Having an error log that results in a no-op seems a bit awkward to me.

Reply:
Won't this error log always trigger for fully filled stateful orders, stateful orders which were filled and then cancelled, stateful orders which were filled and then removed, and stateful orders that were filled and then expired? Also, if we wrongfully trigger this error log we will also not remove the order from the book as a result because we return early.

The `PPME.OrderIdsFilledInLastBlock` does not filter out cancellations, removals, or expirations.

Reply:
>In the future do we want to panic in this situation instead of early returning? Having an error log that results in a no-op seems a bit awkward to me.

Panicing would halt the chain so I avoided it. I think there are cases where a bug with fill amounts could make sense to halt the chain (Short-Term orders not storing ill amounts before expiration could lead to replay attacks), but generally I don't think it makes sense to halt. Let me know if you disagree.

Reply:
> Also, if we wrongfully trigger this error log we will also not remove the order from the book as a result because we return early.

Isn't this just an error log, so it doesn't affect application behavior? It's fine to print out whatever we want as long as we don't panic

> Agreed regarding error cases for stateful orders that were fully-filled, partially-filled then canceled, expired or removed.

Makes sense to me! Agreed that if an order was cancelled or removed this would trigger.

>Panicing would halt the chain so I avoided it.
I think if we have stronger invariants that we are 100% sure must be valid we should halt. In this specific case I think that we shouldn't halt because of the edge cases Jakob said above.

Honestly for this specific function I wouldn't be against calling it `RemoveOrderIfFilled` (which implies it only removes orders if filled, and if not filled, does nothing) and doing all the invariant checking in the PCS callsite.

Reply:
>I think if we have stronger invariants that we are 100% sure must be valid we should halt. In this specific case I think that we shouldn't halt because of the edge cases Jakob said above.

Personally I think it depends on the problem. If the issue doesn't cause traders to lose money (rate-limiting that is too aggressive and leads to dropped orders in `DeliverTx` for example) then I don't think that's worth halting the chain for.

However if it causes traders to lose significant amounts of money, the chain should be halted.

>Honestly for this specific function I wouldn't be against calling it RemoveOrderIfFilled (which implies it only removes orders if filled, and if not filled, does nothing) and doing all the invariant checking in the PCS callsite.

LMK if I'm missing your point, but isn't the function already called `RemoveOrderIfFilled`?

Overall agree about the invariant checking though, I also think that should be done in the PCS callsite especially since that function has the `ProcessProposerMatchesEvents` variables defined. Ideally functions like this return a boolean indicating whether the order was removed or not, which can allow `PurgeInvalidMemclobState` to perform invariant checks

Reply:
> If the issue doesn't cause traders to lose money

Yeah, this school of thought sounds fine to me.

> Isn't the function already called RemoveOrderIfFilled?

Yep, it's already called `RemoveOrderIfFilled`, was commenting on the fact that the fn name lends itself to having invariant checks done in PCS